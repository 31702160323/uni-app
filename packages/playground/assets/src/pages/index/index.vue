<template>
    <view>
        <test-com></test-com>
        <image src="./static/invalid-static.jpg" />
        <image src="../../static/logo.png" />
    </view>
</template>
<script>
    export default {
        onLoad() {
        }
    }
</script>