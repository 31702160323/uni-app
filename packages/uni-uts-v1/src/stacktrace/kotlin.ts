import path from 'path'
import fs from 'fs-extra'
import { relative } from '../utils'
import { originalPositionFor } from '../sourceMap'
import { generateCodeFrame } from './utils'

export interface MessageSourceLocation {
  type: 'exception' | 'error' | 'warning' | 'info' | 'logging' | 'output'
  message: string
  file?: string
  line?: number
  column?: number
  code?: string
}

interface GenerateCodeFrameOptions {
  inputDir: string
  sourceMapDir: string
  replaceTabsWithSpace?: boolean
  format: (msg: MessageSourceLocation) => string
}

export function hbuilderFormatter(m: MessageSourceLocation) {
  const msgs: string[] = []
  let msg = m.type + ': ' + m.message
  if (m.type === 'warning') {
    // 忽略部分警告
    if (msg.includes(`Classpath entry points to a non-existent location:`)) {
      return ''
    }
    msg
      .replace(/\r\n/g, '\n')
      .split('\n')
      .forEach((m) => {
        msgs.push('\u200B' + m + '\u200B')
      })
  } else if (m.type === 'error' || m.type === 'exception') {
    msg
      .replace(/\r\n/g, '\n')
      .split('\n')
      .forEach((m) => {
        msgs.push('\u200C' + m + '\u200C')
      })
  } else {
    msgs.push(msg)
  }
  if (m.file) {
    if (m.file.includes('?')) {
      ;[m.file] = m.file.split('?')
    }
    msgs.push(`at ${m.file}:${m.line}:${m.column}`)
  }
  if (m.code) {
    msgs.push(m.code)
  }
  return msgs.join('\n')
}

export async function parseUTSKotlinStacktrace(
  messages: MessageSourceLocation[],
  options: GenerateCodeFrameOptions
) {
  if (typeof messages === 'string') {
    try {
      messages = JSON.parse(messages)
    } catch (e) {}
  }
  const msgs: string[] = []
  if (Array.isArray(messages) && messages.length) {
    function resolveSourceMapFile(file: string) {
      const sourceMapFile = path.resolve(
        options.sourceMapDir,
        relative(file, options.inputDir) + '.map'
      )
      if (fs.existsSync(sourceMapFile)) {
        return sourceMapFile
      }
    }
    for (const m of messages) {
      if (m.file) {
        const sourceMapFile = resolveSourceMapFile(m.file)
        if (sourceMapFile) {
          const originalPosition = await originalPositionFor({
            sourceMapFile,
            line: m.line!,
            column: m.column!,
            withSourceContent: true,
          })

          if (originalPosition.source && originalPosition.sourceContent) {
            m.file = originalPosition.source.split('?')[0]
            if (originalPosition.line !== null) {
              m.line = originalPosition.line
            }
            if (originalPosition.column !== null) {
              m.column = originalPosition.column
            }
            if (
              originalPosition.line !== null &&
              originalPosition.column !== null
            ) {
              m.code = generateCodeFrame(originalPosition.sourceContent, {
                line: originalPosition.line,
                column: originalPosition.column,
              }).replace(/\t/g, ' ')
            }
          }
        }
      }
      const msg = options.format(m)
      if (msg) {
        msgs.push(msg)
      }
    }
  }
  return msgs.join('\n')
}
