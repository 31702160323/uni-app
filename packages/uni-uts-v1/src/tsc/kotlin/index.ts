import fs from 'fs-extra'
import path from 'path'
import { extend } from '@vue/shared'
import type {
  CompilerOptions,
  SemanticDiagnosticsBuilderProgram,
} from 'typescript'
import {
  type RawSourceMap,
  SourceMapConsumer,
  SourceMapGenerator,
} from 'source-map-js'
import { createBasicUtsOptions } from '../utils/options'
import { isInHBuilderX, normalizePath } from '../../shared'

export interface UTS2KotlinOptions {
  typescript?: typeof import('typescript')
  inputDir: string
  cacheDir: string
  outputDir: string
  tsconfig?: string
  rootFiles?: string[]
  compilerOptions?: CompilerOptions
  normalizeFileName: (str: string) => string
}

export declare class WatchProgramHelper {
  watch(timeout?: number): void
  wait(): Promise<void>
}
export function runUTS2Kotlin(
  mode: 'development' | 'production',
  options: UTS2KotlinOptions
): {
  watcher?: WatchProgramHelper
} {
  const { /* check, noCache, */ tsconfig, typescript, tsconfigOverride } =
    createBasicUtsOptions(options.inputDir)

  const kotlinTypesPath = path.resolve(__dirname, '../../../lib/kotlin/types')
  const rootFiles: string[] = [path.resolve(kotlinTypesPath, 'global.d.ts')]

  ;['env.d.ts', 'shim-uni.d.ts', 'shim-dom.d.ts', 'global.d.ts'].forEach(
    (file) => {
      rootFiles.push(path.resolve(__dirname, '../../../lib/tsconfig', file))
    }
  )
  rootFiles.push(path.resolve(options.inputDir, 'main.uts.ts'))
  const pluginPath = isInHBuilderX()
    ? process.env.UNI_HBUILDERX_PLUGINS
    : path.resolve(process.cwd(), '../')
  const nodeModulesPath = path.resolve(
    pluginPath,
    'hbuilderx-language-services/builtin-dts/uniappx/node_modules'
  )
  const vueRuntimeDts = [
    path.resolve(nodeModulesPath, '@vue/runtime-core/index.d.ts'),
  ]
  extend(tsconfigOverride.compilerOptions.paths, {
    '@dcloudio/uni-runtime': [
      path.resolve(
        kotlinTypesPath,
        '@dcloudio/uni-runtime/dist/uni-runtime.d.ts'
      ),
    ],
    '@vue/reactivity': [
      path.resolve(nodeModulesPath, '@vue/reactivity/dist/reactivity.d.ts'),
    ],
    '@vue/runtime-core': vueRuntimeDts,
    vue: vueRuntimeDts,
  })

  type RunDevOptions = Required<
    UTS2KotlinOptions & {
      sourceMapCallback?: (
        id: string,
        map: string,
        writeFile: (fileName: string, text: string) => void
      ) => boolean | undefined
      callback(
        files: string[],
        program: SemanticDiagnosticsBuilderProgram
      ): void
    }
  >

  return require('../../../lib/kotlin').compile(mode, {
    typescript,
    inputDir: options.inputDir,
    cacheDir: options.cacheDir,
    tsconfig,
    rootFiles,
    compilerOptions: extend(tsconfigOverride.compilerOptions, {
      outDir: options.outputDir,
      inlineSources: true,
      lib: [],
    }),
    normalizeFileName: options.normalizeFileName,
    sourceMapCallback: (fileName, text, writeFile) => {
      const relativeFileName = normalizePath(
        path.relative(options.outputDir, fileName)
      )

      if (fileName.endsWith('.uvue.map') || fileName.endsWith('.vue.map')) {
        const sourceMapFilename = path.resolve(
          options.inputDir,
          relativeFileName.replace('.map', '.ts.map')
        )
        if (fs.existsSync(sourceMapFilename)) {
          // 合并sourcemap
          const sourceMap = fs.readFileSync(sourceMapFilename, 'utf8')
          text = merge(sourceMap, text)
        }
      }
      writeFile(fileName, normalizeSourceMap(text))
      return true
    },
  } as RunDevOptions)
}

function merge(oldMapStr: string, newMapStr: string) {
  if (!oldMapStr) return newMapStr
  if (!newMapStr) return oldMapStr

  const oldMap = JSON.parse(oldMapStr) as RawSourceMap
  const newMap = JSON.parse(newMapStr) as RawSourceMap
  const oldMapConsumer = new SourceMapConsumer(oldMap)
  const newMapConsumer = new SourceMapConsumer(newMap)
  const mergedMapGenerator = new SourceMapGenerator(oldMap)

  // iterate on new map and overwrite original position of new map with one of old map
  newMapConsumer.eachMapping(function (m) {
    // pass when `originalLine` is null.
    // It occurs in case that the node does not have origin in original code.
    if (m.originalLine == null) return

    const origPosInOldMap = oldMapConsumer.originalPositionFor({
      line: m.originalLine,
      column: m.originalColumn,
    })

    if (origPosInOldMap.source == null) return

    mergedMapGenerator.addMapping({
      original: {
        line: origPosInOldMap.line,
        column: origPosInOldMap.column,
      },
      generated: {
        line: m.generatedLine,
        column: m.generatedColumn,
      },
      source: origPosInOldMap.source,
      name: origPosInOldMap.name,
    })
  })

  const maps = [oldMap, newMap]
  maps.forEach(function (map, index) {
    map.sources.forEach(function (sourceFile) {
      const sourceContent = (
        index === 0 ? oldMapConsumer : newMapConsumer
      ).sourceContentFor(sourceFile)
      if (sourceContent != null) {
        mergedMapGenerator.setSourceContent(sourceFile, sourceContent)
      }
    })
  })

  return mergedMapGenerator.toString()
}

function normalizeSourceMap(text: string) {
  const sourceMap = JSON.parse(text) as RawSourceMap
  sourceMap.sources = sourceMap.sources.map((source) => {
    const parts = source.split('/.tsc/')
    source = parts[1] || parts[0]
    source = source.split('?')[0]
    return source.replace('.uts.ts', '.uts')
  })
  return JSON.stringify(sourceMap)
}
