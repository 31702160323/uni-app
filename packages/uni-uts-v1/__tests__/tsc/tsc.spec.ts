import { basename, resolve } from 'path'
import { existsSync, readFileSync, readdirSync } from 'fs-extra'
import { runUTS2Kotlin } from '../../src/tsc/kotlin/index'

function resolvePlatformProjectDirs(platform: 'android') {
  const platformDir = resolve(__dirname, platform)
  const projectDirs: string[] = []
  readdirSync(platformDir).filter((dir) => {
    const androidProjectDir = resolve(platformDir, dir)
    if (existsSync(resolve(androidProjectDir, 'src/index.ts'))) {
      projectDirs.push(androidProjectDir)
    }
  })
  return projectDirs
}

function normalizeFileName(str: string): string {
  return str
}
describe('android', () => {
  resolvePlatformProjectDirs('android').forEach((projectDir) => {
    test(basename(projectDir), () => {
      const outputDir = resolve(projectDir, 'dist')
      runUTS2Kotlin('production', {
        inputDir: resolve(projectDir, 'src'),
        outputDir,
        cacheDir: resolve(projectDir, 'cache'),
        rootFiles: [resolve(projectDir, 'src/index.ts')],
        normalizeFileName,
      })
      expect(
        readFileSync(resolve(outputDir, 'index.uts'), 'utf8')
      ).toMatchSnapshot()
    })
  })
})
