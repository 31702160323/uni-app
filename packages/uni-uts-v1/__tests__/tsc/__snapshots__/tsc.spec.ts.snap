// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`android returnType 1`] = `
"import { createElementVNode, defineComponent } from 'vue';
const __sfc__ = defineComponent({
    methods: {
        num(): number {
            return 1;
        },
        str(): string {
            return '';
        },
        bool(): boolean {
            return true;
        },
    },
});
function render(this: InstanceType<typeof __sfc__>): any | null {
    const _ctx = this;
    return createElementVNode('view', utsMapOf({
        num: (): number => _ctx.num(),
        str: (): string => _ctx.str(),
        bool: (): boolean => _ctx.bool(),
    }));
}
"
`;
